File: summary/ruby/20_bit_dp.rb
def bit_dp(weights, values, capacity)
  n = weights.length
  dp = Array.new(1 << n, 0)

  (1...(1 << n)).each do |s|
    (0...n).each do |i|
      if s & (1 << i) != 0 
        prev_subset = s ^ (1 << i)
        if weights[i] <= capacity
          dp[s] = [dp[s], dp[prev_subset] + values[i]].max
        end
      end
    end
  end

  dp[(1 << n) - 1]
end

# Example usage:
weights = gets.chomp.split.map(&:to_i)
values = gets.chomp.split.map(&:to_i)
capacity = gets.to_i
max_value = bit_dp(weights, values, capacity)
puts "Maximum value using bit DP: #{max_value}"

File: summary/python/20_bit_dp.py
def bit_dp(weights, values, capacity):
    n = len(weights)
    dp = [0] * (1 << n)

    for s in range(1, 1 << n):
        for i in range(n):
            if s & (1 << i):
                prev_subset = s ^ (1 << i)
                if weights[i] <= capacity:
                    dp[s] = max(dp[s], dp[prev_subset] + values[i])

    return dp[(1 << n) - 1]

# Example usage:
weights = list(map(int, input().split()))
values = list(map(int, input().split()))
capacity = int(input())
max_value = bit_dp(weights, values, capacity)
print(f"Maximum value using bit DP: {max_value}")
